docs: added documentation for django (#560)
GoogleCloudPlatform/pgadapter: samples/python/django/README.md

```

You can also drop an existing data model using the `drop_data_model.sql` script:

```shell
psql -h localhost -p 5432 -d my-database -f drop-data-model.sql
```

## Data Types
Cloud Spanner supports the following data types in combination with `Django`.

| PostgreSQL Type                        | Django Model Field                           |
|----------------------------------------|----------------------------------------------|
| boolean                                | BoolField                                    |
| bigint / int8                          | BigIntegerField                              |
| varchar                                | CharField                                    |
| text                                   | CharField                                    |
| float8 / double precision              | FloatField                                   |
| numeric(x, y)                          | DecimalField(max_digits=x, decimal_places=y) |
| timestamptz / timestamp with time zone | DateTimeField                                |
| bytea                                  | BinaryField                                  |

> | date                                   | DateField                                    |

Reviewer: Can we add `jsonb` to this list and to the sample? The `Venue.description` field is intended as the `jsonb` sample field (it's currently `text` in some older samples)

## Limitations
The following limitations are currently known:

| Limitation             | Workaround                                                                                                                                                                                                                                                                                          |
|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Annotate               | Cloud Spanner does not support selecting columns in an aggregation query that are not involved in the aggregation functions or `GROUP BY` clause. Due to this, many times performing `annotate` function over a querySet might fail.                                                                |

Reviewer: ```suggestion
| Annotate               | Cloud Spanner does not support selecting columns in an aggregation query that are not included in the aggregation functions or `GROUP BY` clause. Adding an aggregation on a query using `annotate` will therefore often generate an unsupported query.                                                                |
```

| Generated primary keys | Cloud Spanner does not support `sequences`. Auto-increment primary key is not supported. Avoid using fields like `AutoField`, `BigAutoField`, etc. . The recommended type of primary key is a client side generated `UUID` stored as a string.                                                      |

> | Savepoint              | Cloud Spanner does not support `savepoints` and hence, functions related to `savepoints` in django are not supported in the PGAdapter.                                                                                                                                                              |

Reviewer: ```suggestion
| Savepoint              | Cloud Spanner does not support `savepoints`. Creating or rolling back to a `Savepoint` in django is not supported in PGAdapter.                                                                                                                                                              |
```

> | Nested Atomic Blocks   | Nested atomic blocks make use of `savepoints`, which is currently not supported by Cloud Spanner. Hence, while using nested atomic blocks make sure to set `savepoint=False`, in the argument.                                                                                                      |

Reviewer: ```suggestion
| Nested Atomic Blocks   | Nested atomic blocks use `savepoints`, which is currently not supported by Cloud Spanner. Ensure to set `savepoint=False` when using nested atomic blocks.                                                                                                      |
```

> | Interleaved Table      | Django doesn't support composite primary keys and interleaved table in cloud spanner requires a composite primary key and on the top of that, it is also not possible to create a django model without a primary key. A possible workaround around for using interleaved table is documented below. |

Reviewer: ```suggestion
| Interleaved Table      | Django does not support composite primary keys. Interleaved tables in Cloud Spanner require composite primary keys. A possible workaround for using interleaved tables is documented below. |
```

### Annotate

> `annotate` is not fully supported in the PGAdapter.

Reviewer: Add more context and a concrete example for something that does not work. And if there's a workaround, then add an example for that as well.

### Generated Primary Keys
`Sequences` are not supported. Hence, auto increment primary key is not supported and should be replaced with primary key definitions that
are manually assigned. See https://cloud.google.com/spanner/docs/schema-design#primary-key-prevent-hotspots
for more information on choosing a good primary key. This sample uses UUIDs that are generated by the client for primary
keys.

### Savepoint

> `savepoints` are supported in the Cloud Spanner. Hence, following functions from the `django.db.transaction` are not supported

Reviewer: ```suggestion
`savepoints` are not supported in Cloud Spanner. The following functions in Django are therefore also not supported:
```

```python
from django.db import  transaction
transaction.savepoint()
transaction.savepoint_rollback()
transaction.savepoint_commit()
transaction.clean_savepoints()
```

### Nested Atomic Blocks
As mentioned earlier also, `savepoints` are not supported and hence, a code like follows might not work with PGAdapter
because it will make use of savepoint to make sure the inner nested block is rolled back properly 

> if some error occurs over there.

Reviewer: This section is too much like 'informal spoken language'. Please rephrase to:

```suggestion
`Savepoints` are not supported in Cloud Spanner. Nested atomic blocks in `django` will by default create a
`Savepoint` when an inner block is entered. You can suppress this behavior by adding `savepoint=False` to
the `@transaction.atomic` annotation.
```

```python
from django.db import transaction

@transaction.atomic
def func():
  do_stuff
  with transaction.atomic():
    do_some_more_stuff
```

> Instead of above, always set `savepoint=False` while using nested atomic blocks

Reviewer: Remove this and only show the sample that works

```python
from django.db import transaction

> @transaction.atomic(savepoint=False)

Reviewer: Add a code comment here and below to indicate why we are including `savepoint=False`, e.g.:

`# Include 'savepoint=False' to prevent django from creating a savepoint.`

def func():
  do_stuff
  with transaction.atomic(savepoint=False):
    do_some_more_stuff
```

### Interleaved Tables
A possible workaround for Django's inability to use interleaved table with Django and PGAdapter includes 
having a column which has unique index in the actual table while pretending to Django like it's a primary key. 

One important thing to note here is that we can't use usual `save` function for creating new records for such a model 
because that might generate an `UPDATE` statement which will try to update the unique index. 

> So, we will have to use `force_insert = True`, while creating new records. For more details, see the sample application.

Reviewer: [Interleaved tables](https://cloud.google.com/spanner/docs/schema-and-data-model#primary-keys) in Cloud Spanner require composite primary keys. Django does not support models with a composite primary key. You
can work around these limitations and use an interleaved table with Cloud Spanner and Django as follows:

1. Manually create the interleaved table in your database by executing a `CREATE TABLE` statement. This table will use a composite primary key.
2. Define the model in Python code so that the parent key column of the table is defined as a `ForeignKey` referencing the parent table, and the child key column is defined as the primary key of the table.
3. Add a unique index to the child primary key column. This will ensure that the child primary key part is always globally unique, and can be used to efficiently look up a single row.

Example:

```sql
CREATE TABLE tracks (
    track_id character varying NOT NULL,
    id character varying NOT NULL,
    ...
    PRIMARY KEY(id, track_number)
) INTERLEAVE IN PARENT albums ON DELETE CASCADE;

CREATE UNIQUE INDEX unique_idx_track_id ON tracks(track_id);
```

```python
class Track(BaseModel):
  class Meta():
    db_table = 'tracks'
  track_id = models.CharField(primary_key=True, null=False)
  album = models.ForeignKey(Album, on_delete=models.DO_NOTHING, db_column='id')
```

Note that using the standard `save` function in Django is not supported for interleaved tables, as this function will also try to update the `album` column. This column is part of the primary key definition of the table and cannot be updated. Add a `force_insert=True` to create new child records:

```python
track.save(force_insert=True)
```

See the sample application for more details and a working example.

  -x
```

## Creating the Sample Data Model
Run the following command in this directory. Replace the host, port and database name with the actual host, port and
database name for your PGAdapter and database setup.

```shell
psql -h localhost -p 5432 -d my-database -f sample-schema.sql
```

You can also drop an existing data model using the `drop_data_model.sql` script:

```shell
psql -h localhost -p 5432 -d my-database -f drop-data-model.sql
```

## Data Types
Cloud Spanner supports the following data types in combination with `Django`.

| PostgreSQL Type                        | Django Model Field                           |
|----------------------------------------|----------------------------------------------|
| boolean                                | BoolField                                    |
| bigint / int8                          | BigIntegerField                              |
| varchar                                | CharField                                    |
| text                                   | CharField                                    |
| float8 / double precision              | FloatField                                   |

> | numeric(x, y)                          | DecimalField(max_digits=x, decimal_places=y) |

Reviewer: nit: Cloud Spanner does not support type modifiers for `numeric`. That is; only `numeric` and not `numeric(x, y)` is supported.

Reviewer: ping: this still needs to be changed

| timestamptz / timestamp with time zone | DateTimeField                                |
| bytea                                  | BinaryField                                  |
| date                                   | DateField                                    |

## Limitations
The following limitations are currently known:

> | Limitation                           | Workaround                                                                                                                                         |

Reviewer: This table should at least also include:
1. The limitation that interleaved tables are not fully supported, other than with some workarounds that come with its own limitations (including the lmitations around saving / updating rows in an interleaved table).
2. Nested transactions / savepoints.

| Generated primary keys               | Cloud Spanner does not support `sequences`. Auto-increment primary key is not supported. Remove auto increment annotation for primary key columns. The recommended type of primary key is a client side generated `UUID` stored as a string. |

### Schema Updates
Schema updates are not supported as Cloud Spanner does not support the full PostgreSQL DDL dialect. It is recommended to
create the schema manually. Note that PGAdapter does support `create table if not exists` / `drop table if exists`.
See [sample-schema.sql](src/main/resources/sample-schema-sql) for the data model for this example.

### Generated Primary Keys
`Sequences` are not supported. Hence, auto increment primary key is not supported and should be replaced with primary key definitions that
are manually assigned. See https://cloud.google.com/spanner/docs/schema-design#primary-key-prevent-hotspots
for more information on choosing a good primary key. This sample uses UUIDs that are generated by the client for primary
keys.

```java

> public class User {

Reviewer: Replace with a Python example

	// This uses auto generated UUID.
    @Id
    @Column(columnDefinition = "varchar", nullable = false)
    @GeneratedValue
    private UUID id;

> }

Reviewer: We should add a 'Performance Considerations' section for Django as well. See https://github.com/GoogleCloudPlatform/pgadapter/blob/e7c4797ce35cd0fccc688a1269731762cd1c9c86/samples/python/sqlalchemy-sample/README.md#performance-considerations for an example.
